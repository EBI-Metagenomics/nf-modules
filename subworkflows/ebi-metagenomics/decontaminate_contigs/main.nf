include { MINIMAP2_ALIGN } from '../../../modules/ebi-metagenomics/minimap2/align/main'
include { FILTERPAF      } from '../../../modules/ebi-metagenomics/filterpaf/main'
include { SEQKIT_GREP    } from '../../../modules/ebi-metagenomics/seqkit/grep/main'

workflow DECONTAMINATE_CONTIGS {
    /*
    * Important: meta compatibility requirement
    *
    * When using this subowkrflow, ensure that 'contigs' and 'contaminant_genomes'
    * channels are joinable by the first element of their meta attributes.
    *
    * This is required to properly join the channels and prevent cleaning contigs
    * with incorrect reference genomes.
    */

    // ================================================================================
    // Decontamination
    //  Remove contigs with alignment > min_align_len bp
    //    OR
    //  (query coverage ≥ min_qcov AND mapping quality ≥ min_mapq)
    //
    // Rationale: Long PacBio HiFi contigs (8+ Mbp) may have large partial
    // alignments (3+ Mbp) to contaminants, making absolute length more
    // informative than coverage percentage.
    // ================================================================================

    take:
    contigs             // [ meta, path(assembly_fasta)]
    contaminant_genomes // [ meta, path(reference_fasta)]

    main:
    ch_versions = Channel.empty()

    // Mix and match the genomes and contigs
    contigs
        .join(contaminant_genomes)
        .multiMap { meta, assembly, cont_genome ->
            contigs: [meta, assembly]
            genome: [meta, cont_genome]
        }
        .set { synced_contigs_contaminant_genome }

    MINIMAP2_ALIGN(
        synced_contigs_contaminant_genome.contigs,
        synced_contigs_contaminant_genome.genome,
        false,             // bam_format
        false,             // bam_index_extension
        false,             // cigar_paf_format
        false              // cigar_bam
    )
    ch_versions = ch_versions.mix(MINIMAP2_ALIGN.out.versions.first())

    FILTERPAF(
        MINIMAP2_ALIGN.out.paf
    )
    ch_versions = ch_versions.mix(FILTERPAF.out.versions.first())

    contigs
        .join(FILTERPAF.out.mapped_contigs_txt)
        .multiMap { meta, assembly, mapped_contigs_txt ->
            contigs: [meta, assembly]
            pattern: mapped_contigs_txt
        }
        .set { seqkit_grep_input_ch }

    SEQKIT_GREP(
        seqkit_grep_input_ch.contigs,
        seqkit_grep_input_ch.pattern,
    )
    ch_versions = ch_versions.mix(SEQKIT_GREP.out.versions.first())

    emit:
    cleaned_contigs = SEQKIT_GREP.out.filter
    versions        = ch_versions
}
