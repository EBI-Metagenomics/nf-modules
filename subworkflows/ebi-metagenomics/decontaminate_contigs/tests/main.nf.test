nextflow_workflow {

    name "Test subworkflow DECONTAMINATE_CONTIGS"
    script "../main.nf"
    config "../nextflow.config"
    workflow "DECONTAMINATE_CONTIGS"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/decontaminate_contigs"
    tag "minimap2/align"
    tag "seqkit/grep"
    tag "filterpaf"

    test("Run DECONTAMINATE_CONTIGS swf - contaminated contigs input") {
        when {
            params {
                output_path = "${outputDir}"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ [id:'test'], file("${moduleDir}/tests/data/test_contaminated.fa", checkIfExists: true) ],
                        file("${moduleDir}/tests/data/ref.fa", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                {
                    def contaminated_contigs = path("${moduleDir}/tests/data/test_contaminated.fa").fasta
                    assert contaminated_contigs.size() == 4
                    assert "ERZ18545498.1194 atg000172l" in contaminated_contigs
                    assert "ERZ18545498.8 ptg000008l" in contaminated_contigs
                    assert "ERZ18545498.980 ptg000980l" in contaminated_contigs
                    assert "Fake contamination" in contaminated_contigs
                },
                {
                    def cleaned_contigs = path(workflow.out.cleaned_contigs.get(0).get(1)).fasta
                    assert "ERZ18545498.1194 atg000172l" in cleaned_contigs
                    assert "ERZ18545498.8 ptg000008l" in cleaned_contigs
                    // Contaminated contigs
                    assert !("ERZ18545498.980 ptg000980l" in cleaned_contigs)
                    assert !("Fake contamination" in cleaned_contigs)
                },
            )
        }
    }

    test("Run DECONTAMINATE_CONTIGS swf - contaminated contigs input gz") {
        when {
            params {
                output_path = "${outputDir}"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ [id:'test'], file("${moduleDir}/tests/data/test_contaminated.fa.gz", checkIfExists: true) ],
                        file("${moduleDir}/tests/data/ref.fa", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                {
                    def contaminated_contigs = path("${moduleDir}/tests/data/test_contaminated.fa.gz").fasta
                    assert contaminated_contigs.size() == 4
                    assert "ERZ18545498.1194 atg000172l" in contaminated_contigs
                    assert "ERZ18545498.8 ptg000008l" in contaminated_contigs
                    assert "ERZ18545498.980 ptg000980l" in contaminated_contigs
                    assert "Fake contamination" in contaminated_contigs
                },
                {
                    def cleaned_contigs = path(workflow.out.cleaned_contigs.get(0).get(1)).fasta
                    assert "ERZ18545498.1194 atg000172l" in cleaned_contigs
                    assert "ERZ18545498.8 ptg000008l" in cleaned_contigs
                    // Contaminated contigs
                    assert !("ERZ18545498.980 ptg000980l" in cleaned_contigs)
                    assert !("Fake contamination" in cleaned_contigs)
                },
            )
        }
    }

    test("Run DECONTAMINATE_CONTIGS swf - uncontaminated contigs input") {
        when {
            params {
                output_path = "${outputDir}"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ [id:'test'], file("${moduleDir}/tests/data/test_not_contaminated.fa", checkIfExists: true) ],
                        file("${moduleDir}/tests/data/ref.fa", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                {
                    def not_contaminated_contigs = path("${moduleDir}/tests/data/test_not_contaminated.fa").fasta
                    assert not_contaminated_contigs.size() == 2
                    assert "ERZ18545498.1194 atg000172l" in not_contaminated_contigs
                    assert "ERZ18545498.8 ptg000008l" in not_contaminated_contigs
                },
                {
                    def cleaned_contigs = path(workflow.out.cleaned_contigs.get(0).get(1)).fasta
                    assert "ERZ18545498.1194 atg000172l" in cleaned_contigs
                    assert "ERZ18545498.8 ptg000008l" in cleaned_contigs
                    assert cleaned_contigs.size() == 2
                },
            )
        }
    }

    test("Run DECONTAMINATE_CONTIGS swf - uncontaminated contigs input gz") {
        when {
            params {
                output_path = "${outputDir}"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ [id:'test'], file("${moduleDir}/tests/data/test_not_contaminated.fa.gz", checkIfExists: true) ],
                        file("${moduleDir}/tests/data/ref.fa", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                {
                    def not_contaminated_contigs = path("${moduleDir}/tests/data/test_not_contaminated.fa.gz").fasta
                    assert not_contaminated_contigs.size() == 2
                    assert "ERZ18545498.1194 atg000172l" in not_contaminated_contigs
                    assert "ERZ18545498.8 ptg000008l" in not_contaminated_contigs
                },
                {
                    def cleaned_contigs = path(workflow.out.cleaned_contigs.get(0).get(1)).fasta
                    assert "ERZ18545498.1194 atg000172l" in cleaned_contigs
                    assert "ERZ18545498.8 ptg000008l" in cleaned_contigs
                    assert cleaned_contigs.size() == 2
                },
            )
        }
    }
}
