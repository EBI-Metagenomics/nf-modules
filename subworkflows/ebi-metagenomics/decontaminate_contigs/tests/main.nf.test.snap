{
    "Run DECONTAMINATE_CONTIGS swf - contaminated contigs input gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "1": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ],
                "cleaned_contigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T09:54:13.174391"
    },
    "Run DECONTAMINATE_CONTIGS swf - uncontaminated contigs input": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "1": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ],
                "cleaned_contigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T10:17:05.214974"
    },
    "Run DECONTAMINATE_CONTIGS swf - uncontaminated contigs input gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "1": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ],
                "cleaned_contigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T10:20:59.35183"
    },
    "Run DECONTAMINATE_CONTIGS swf - contaminated contigs input": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "1": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ],
                "cleaned_contigs": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cleaned.fa.gz:md5,2e05937059d9329d7b8d56189c9d8c10"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,168ddf5d2325c3dfe5959865ead422c3",
                    "versions.yml:md5,a29b0abb3251c925c794dd3d01a9e47b",
                    "versions.yml:md5,fbcc2bf4987f9f05018d9ffa49cb2d98"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T09:54:09.174981"
    }
}